lib/Runtime/Library/JavascriptBigInt.cpp

```diff
@@ -113,26 +156,182 @@ namespace Js
+                AssertOrFailFastMsg(false, "AbsoluteIncrement overflow");
+            }
+            result->m_digits[result->m_length++] = carry;
+        }
+        return result;
+    }
+
+    JavascriptBigInt * JavascriptBigInt::AbsoluteDecrement(JavascriptBigInt * pbi)
+    {
+        JavascriptBigInt* result = pbi;
+        Assert(!JavascriptBigInt::IsZero(result));
+        digit_t borrow = 1;
+        for (digit_t i = 0; i < result->m_length && borrow > 0; i++)
+        {
+            digit_t tempBorrow = 0;
+            result->m_digits[i] = JavascriptBigInt::SubDigit(result->m_digits[i], borrow, &tempBorrow);
+            borrow = tempBorrow;
+        }
+        Assert(borrow == 0);
+        // remove trailing zero
+        if (result->m_digits[result->m_length-1] == 0)
+        {
+            result->m_length--;
+        }
+        return result;
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Increment(JavascriptBigInt * pbi)
+    {
+        if (pbi->m_isNegative)
+        {
+            // return 0n for -1n
+            if (pbi->m_length == 1 && pbi->m_digits[0] == 1)
+            {
+                JavascriptBigInt* result = pbi;
+                result->m_digits[0] = 0;
+                result->m_isNegative = false;
+                return result;
+            }
+            return JavascriptBigInt::AbsoluteDecrement(pbi);
+        }
+        return JavascriptBigInt::AbsoluteIncrement(pbi);
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Decrement(JavascriptBigInt * pbi)
+    {
+        if (pbi->m_isNegative)
+        {
+            return JavascriptBigInt::AbsoluteIncrement(pbi);
+        }
+        if (JavascriptBigInt::IsZero(pbi)) // return -1n for 0n
+        {
+            JavascriptBigInt* result = pbi;
+            result->m_digits[0] = 1;
+            result->m_isNegative = true;
+            return result;
+        }
+        return JavascriptBigInt::AbsoluteDecrement(pbi);
+    }
+
+    Var JavascriptBigInt::Increment(Var aRight)
+    {
+        AssertMsg(VarIs<JavascriptBigInt>(aRight), "BigInt Increment");
+
+        JavascriptBigInt* rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Increment(rightBigInt);
+    }
+
+    Var JavascriptBigInt::Decrement(Var aRight)
+    {
+        AssertMsg(VarIs<JavascriptBigInt>(aRight), "BigInt Increment");
+
+        JavascriptBigInt* rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Decrement(rightBigInt);
+    }
+
+    // return low(a*b) and out high
+    digit_t JavascriptBigInt::MulDigit(digit_t a, digit_t b, digit_t* resultHigh)
+    {
+        // Multiply is performed in half chuck favor.
+        // For inputs [AH AL]*[BH BL], the result is:
+        //
+        //            [AL*BL]  // rLow
+        //    +    [AL*BH]     // rMid1
+        //    +    [AH*BL]     // rMid2
+        //    + [AH*BH]        // rHigh
+        //    = [R1 R2 R3 R4]  // high = [R1 R2], low = [R3 R4]
+        //
+#ifdef TARGET_64
```

Comment:
You can avoid the #if stuff by using sizeof. These expressions are compile-time constant so the compiler should do the right thing and emit numbers directly.

```c++
digit_t kHalfDigitBits = sizeof(digit_t) * 4;
digit_t kHalfDigitMask = (1 << kHalfDigitBits) - 1;
```

Reply:
somehow this does not work

Reply:
I add one cast to the 2nd line and it seems to work fine 

`digit_t kHalfDigitBits = sizeof(digit_t) * 4;
digit_t kHalfDigitMask = ((digit_t)1 << kHalfDigitBits) - 1;`